src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs

```diff
@@ -641,55 +666,94 @@ protected override void BeginProcessing()
             _selectObjectQueue = new SelectObjectQueue(_first, _last, Skip, SkipLast, _firstOrLastSpecified);
         }
 
-        private int _indexOfCurrentObject = 0;
-        private int _indexCount = 0;
         /// <summary>
+        /// Handles processing of InputObject.
         /// </summary>
         protected override void ProcessRecord()
         {
             if (InputObject != AutomationNull.Value && InputObject != null)
             {
-                if (!_indexSpecified)
+                if (_indexSpecified)
+                {
+                    ProcessIndexed();
+                }
+                else
                 {
                     _selectObjectQueue.Enqueue(InputObject);
                     PSObject streamingInputObject = _selectObjectQueue.StreamingDequeue();
                     if (streamingInputObject != null)
                     {
                         ProcessObjectAndHandleErrors(streamingInputObject);
                     }
+
                     if (_selectObjectQueue.AllRequestedObjectsProcessed && !this.Wait)
                     {
                         this.EndProcessing();
                         throw new StopUpstreamCommandsException(this);
                     }
                 }
-                else
+            }
+        }
+
+        private int _indexOfCurrentObject;
```

Comment:
Suggestion: 
To me these field names are confusing. _indexOfCurrentObject implies the object being processed but is actually the index into the array of index filters while _indexCount is the index of the object being processed.

I think it would improve readability if you consider changing these names to something like the following, or some variation:
_currentObjectIndex - the index of the object being processed.
_currentFilterIndex - the index of the active index filter.